olive/resource_path.py

```diff
@@ -0,0 +1,387 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import inspect
+import logging
+import shutil
+import tempfile
+from abc import abstractmethod
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Type, Union
+
+from pydantic import Field, validator
+
+from olive.azureml.azureml_client import AzureMLClientConfig
+from olive.common.config_utils import ConfigBase, validate_config
+from olive.common.utils import retry_func
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceType(str, Enum):
+    LocalFile = "file"
+    LocalFolder = "folder"
+    StringName = "string_name"
+    AzureMLModel = "azureml_model"
+    AzureMLDatastore = "azureml_datastore"
+    AzureMLJobOutput = "azureml_job_output"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+LOCAL_RESOURCE_TYPES = [ResourceType.LocalFile, ResourceType.LocalFolder, ResourceType.StringName]
+AZUREML_RESOURCE_TYPES = [ResourceType.AzureMLModel, ResourceType.AzureMLDatastore, ResourceType.AzureMLJobOutput]
+
+_resource_config_registry: Dict[ResourceType, Type[ConfigBase]] = {}
+
+
+class ResourceConfig(ConfigBase):
+    _type: ResourceType = None
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs) -> None:
+        """Register the resource config."""
+        super().__init_subclass__(**kwargs)
+        if not inspect.isabstract(cls) and cls._type is not None:
+            _resource_config_registry[cls._type] = cls
+
+    @abstractmethod
+    def get_path(self) -> str:
+        """Return the resource path as a string."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def save_to_dir(self, dir_path: Union[Path, str], name: str = None, overwrite: bool = False) -> str:
```

Comment:
force better?

Reply:
I think `overwrite` describes the behavior better?